src/write_gcm.c

```diff
@@ -2275,59 +2275,136 @@ static monitored_resource_t *monitored_resource_create_from_fields(
+    .processing_labels = 0
+  };
+
   {
-    const char * type_path[] = {"type", (const char *) 0};
-    const char * labels_path[] = {"labels", (const char *) 0};
-    char *type;
-    yajl_val type_v = yajl_tree_get(node, type_path, yajl_t_string);
-    yajl_val labels_v = yajl_tree_get(node, labels_path, yajl_t_object);
-    if (type_v && YAJL_IS_STRING(type_v)) {
-      type = YAJL_GET_STRING(type_v);
-    } else {
-      ERROR("write_gcm: wg_parse_monitored_resource: %s not correctly defined.",
-            type_path[0]);
+    yajl_callbacks callbacks = {
+      .yajl_null = NULL,
+      .yajl_boolean = NULL,
+      .yajl_integer = NULL,
+      .yajl_double = NULL,
+      .yajl_number = NULL,
+      .yajl_string = &wg_mr_handle_string,
+      .yajl_start_map = NULL,
+      .yajl_map_key = &wg_mr_handle_map_key,
+      .yajl_end_map = &wg_mr_handle_end_map,
+      .yajl_start_array = NULL,
+      .yajl_end_array = NULL
+    };
+
+  #if defined(YAJL_MAJOR) && YAJL_MAJOR >= 2
+    handle = yajl_alloc(&callbacks, NULL, &ctx);
+  #else
+    yajl_parser_config config = { 0, 1 };
+    handle = yajl_alloc(&callbacks, &config, NULL, &ctx);
+  #endif
+
+    if (yajl_parse(handle, (const unsigned char *) metadata, strlen(metadata))
+        != yajl_status_ok) {
+      ERROR("write_gcm: parse_monitored_resource: Invalid JSON response.");
       goto finish;
     }
-    size_t num_values;
-    if (labels_v && YAJL_IS_OBJECT(labels_v)) {
-      num_values = labels_v->u.object.len;
-    } else {
-      ERROR("write_gcm: wg_parse_monitored_resource: No such node %s.\n",
-            labels_path[0]);
+
+    int parse_result;
+  #if defined(YAJL_MAJOR) && YAJL_MAJOR >= 2
+    parse_result = yajl_complete_parse(handle);
+  #else
+    parse_result = yajl_parse_complete(handle);
+  #endif
+
+    if (parse_result != yajl_status_ok) {
+      ERROR("write_gcm: parse_monitored_resource: Invalid JSON response.");
       goto finish;
     }
-    label_t labels[num_values];
-    for (int i = 0; i < num_values; i++) {
-      labels[i].key = *(labels_v->u.object.keys + i);
-      labels[i].value = YAJL_GET_STRING(*(labels_v->u.object.values + i));
+
+    int total_labels = llist_size(ctx.label_list);
+    label_t labels[total_labels];
+
+    llentry_t *curr = llist_head(ctx.label_list);
+    for (int i = 0; curr != NULL; curr = curr->next, ++i) {
+      labels[i].key = curr->key;
+      labels[i].value = (char *) curr->value;
     }
-    response = monitored_resource_create_from_array(type, project_id, labels,
-                   num_values);
+
+    response = monitored_resource_create_from_array(ctx.type, project_id, 
+                                                    labels, total_labels);
   }
 
  finish:
-  yajl_tree_free(node);
+  sfree(ctx.type);
+  llentry_t *curr = llist_head(ctx.label_list);
+  for (int i = 0; curr != NULL; curr = curr->next, ++i) {
+    sfree(curr->key);
+    sfree(curr->value);
+  }
+  llist_destroy(ctx.label_list);
+  yajl_free(handle);
```

Comment:
My 2¢: let's leave this cleanup code as-is -- it's pretty clear what it does, and makes it immediately obvious what gets freed. It might help to have a comment above the loop saying that the keys and values in the list were allocated during JSON parsing.